src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,340 @@
+        }
+
+        var orderedByLowestScore = chooseFromSilos.OrderBy(kv => kv.Value);
+
+        // there could be more than 1 silo that has the same score, we pick 1 of them randomly so that we dont continuously pick the first one.
+        var lowestScore = orderedByLowestScore.First().Value;
+        var shortListedSilos = orderedByLowestScore.TakeWhile(p => p.Value == lowestScore).ToList();
+        var winningSilo = shortListedSilos[Random.Shared.Next(shortListedSilos.Count)];
+
+        return winningSilo;
+    }
+
+    private KeyValuePair<SiloAddress, float> GetBestSiloCandidate_V2(SiloAddress[] compatibleSilos)
+    {
+        (int Index, float Score) pick;
+        int compatibleSilosCount = compatibleSilos.Length;
+
+        // It is good practice not to allocate more than 1 kilobyte of memory on the stack
+        if (compatibleSilosCount * Unsafe.SizeOf<(int, ResourceStatistics)>() <= OneKiloByte)
+        {
+            pick = MakePick(stackalloc (int, ResourceStatistics)[compatibleSilosCount]);
+        }
+        else
+        {
+            var relevantSilos = ArrayPool<(int, ResourceStatistics)>.Shared.Rent(compatibleSilosCount);
+            pick = MakePick(relevantSilos.AsSpan());
+            ArrayPool<(int, ResourceStatistics)>.Shared.Return(relevantSilos);
+        }
+
+        return new KeyValuePair<SiloAddress, float>(compatibleSilos[pick.Index], pick.Score);
+
+        (int, float) MakePick(Span<(int, ResourceStatistics)> relevantSilos)
+        {
+            // Get all compatible silos
+            int relevantSilosCount = 0;
+            for (var i = 0; i < compatibleSilos.Length; ++i)
+            {
+                var silo = compatibleSilos[i];
+                if (_siloStatistics.TryGetValue(silo, out var stats) && !stats.IsOverloaded)
+                {
+                    relevantSilos[relevantSilosCount++] = new(i, stats);
+                }
+            }
+
+            // Limit to the number of candidates added.
+            relevantSilos = relevantSilos[0..relevantSilosCount];
+            Debug.Assert(relevantSilos.Length == relevantSilosCount);
+
+            // Pick K silos from the list of compatible silos, where K is equal to the square root of the number of silos.
+            // Eg, from 10 silos, we choose from 4.
+            int candidateCount = (int)Math.Ceiling(Math.Sqrt(relevantSilosCount));
+            ShufflePrefix(relevantSilos, candidateCount);
+            var candidates = relevantSilos[0..candidateCount];
+
+            (int Index, float Score) pick = (0, 1f);
+
+            foreach (var (index, statistics) in candidates)
+            {
+                float score = CalculateScore(statistics);
+
+                // It's very unlikely, but there could be more than 1 silo that has the same score,
+                // so we apply some jittering to avoid pick the first one in the short-list.
+                float scoreJitter = Random.Shared.NextSingle() / 100_000f;
```

Comment:
There is another reason to add some noise here: it can be used to reduce swarming (more relevant for clusters of ~10 or less, where 40%+ of silos are being examined)
For that reason, I wonder if we should use a higher jitter (eg, 1% jitter) instead of .001% jitter.

Reply:
I am actually feeling a touch uncomfortable even with amount of jittering, increasing it *may* put a risk of the jitter effect overshoot above another silo's actual value which would result in a incorrectly picked target. I would rather remove the jittering altogether, because the PowK runs before this set, so whatever K silos are pick, the chances of them being totally equally in the computed load is very small, add to that the fact the PowK will highly likely pick another set of K silos which weren't on the previous set, and the chances goes even further down.

Reply:
I am ok with that, too. I am not as worried about jittering even couple of %

Reply:
For reference: we ended up keeping the jittering, but still a small amount